src/event-manager/index.ts

```diff
@@ -0,0 +1,50 @@
+import Ain from '../ain';
+import {
+  BlockFinalizedEventConfig,
+  EventChannelConnectionOption,
+  EventConfigType,
+  TxStateChangedEventConfig,
+  ValueChangedEventConfig,
+} from '../types';
+import EventChannelClient from './event-channel-client';
+import EventCallbackManager from './event-callback-manager';
+import Subscription from './subscription';
+
+export default class EventManager {
+  private _ain: Ain;
+  private readonly _eventCallbackManager: EventCallbackManager;
+  private readonly _eventChannelClient: EventChannelClient;
+
+  constructor(ain: Ain) {
+    this._ain = ain;
+    this._eventCallbackManager = new EventCallbackManager();
+    this._eventChannelClient = new EventChannelClient(ain, this._eventCallbackManager);
+  }
+
+  async connect(connectionOption?: EventChannelConnectionOption) {
+    await this._eventChannelClient.connect(connectionOption || {});
+  }
+
+  disconnect() {
+    this._eventChannelClient.disconnect();
+  }
+
+  subscribe(
+      eventType: 'BLOCK_FINALIZED',
+      config: BlockFinalizedEventConfig): Subscription;
+  subscribe(
+      eventType: 'VALUE_CHANGED',
+      config: ValueChangedEventConfig): Subscription;
+  subscribe(
+      eventType: 'TX_STATE_CHANGED',
+      config: TxStateChangedEventConfig): Subscription;
+  subscribe(eventTypeStr: string, config: EventConfigType): Subscription {
```

Comment:
could we add a callback function here directly? e.g.
```
  subscribe(eventTypeStr: string, config: EventConfigType, callback: Function): Subscription {
```

Reply:
question: What is a callback function used for? If it's for event listening, we can use a pattern like the one below!
```
  const subscription = ain.em.subscribe('BLOCK_FINALIZED', {
    block_number: 1000,
  }).on('data', (data) => {

  }).on('error', (error) => {

  });
```

Reply:
it looks like the subscription supports two types of event, 'data' and 'error' for one filter.
can't the subscribe() function handle the details for users? e.g.
```
 subscribe(..., dataCallback: Function, errorCallback: Function): Subscription {
   ...
   return this._eventCallbackManager.createSubscription(filter)
       .on('data', dataCallback)
       .on('error', errorCallback);
 }
 ```
 ?

Reply:
(and now the callback manager is not handling any callbacks?)

Reply:
Looks good. How about providing callback functions as optional? 
```
subscribe(
      eventTypeStr: string, config: EventConfigType, dataCallback?: (data: any) => void,
      errorCallback?: (error: any) => void): Subscription {
    if (!this._eventChannelClient.isConnected) {
      throw Error(`Event channel is not connected! You must call ain.eh.connect() before using subscribe()`);
    }
    const filter = this._eventCallbackManager.createFilter(eventTypeStr, config);
    this._eventChannelClient.registerFilter(filter);
    const subscription = this._eventCallbackManager.createSubscription(filter);
    if (dataCallback) {
      subscription.on('data', dataCallback);
    }
    if (errorCallback) {
      subscription.on('error', errorCallback);
    }
    return subscription;
  }
```

Reply:
Users can set callback functions when call `subscribe()`, and also can access it later through `subscription`.

Reply:
PTAL! @platfowner

Reply:
what about passing the callback function to the event callback manager when a subscription is created? callback handling seems to be the callback manager's responsibility.

Reply:
I agree with your suggestion! Done